pkg/cloudevents/types/value.go

```diff
@@ -10,201 +10,212 @@ import (
-		f, err := strconv.ParseFloat(x, 64)
-		if f > math.MaxInt32 || f < math.MinInt32 {
-			return 0, rangeErr(x)
-		}
-		return int32(f), err
+		return ParseInteger(v)
 	default:
-		return 0, typeErr("Integer", x)
+		return 0, convertErr("Integer", v, err)
 	}
 }
 
-// ToString returns a string value. It does NOT encode non-string, use String() for that.
-func (v Value) ToString() (string, error) {
-	switch x := v.v.(type) {
+// ToString returns a string value unaltered.
+//
+// This function does not perform canonical string encoding, use one of the
+// Format functions for that.
+func ToString(v interface{}) (string, error) {
+	v, err := Validate(v)
+	switch v := v.(type) {
 	case string:
-		return x, nil
+		return v, nil
 	default:
-		return "", typeErr("String", x)
+		return "", convertErr("String", v, err)
 	}
 }
 
-// ToBinary returns a []byte value, decoding from string if necessary.
-func (v Value) ToBinary() ([]byte, error) {
-	switch x := v.v.(type) {
+// ToBinary returns a []byte value, decoding from base64 string if necessary.
+func ToBinary(v interface{}) ([]byte, error) {
+	v, err := Validate(v)
+	switch v := v.(type) {
 	case []byte:
-		return x, nil
+		return v, nil
 	case string:
-		return base64.StdEncoding.DecodeString(x)
+		return base64.StdEncoding.DecodeString(v)
 	default:
-		return nil, typeErr("Binary", x)
+		return nil, convertErr("Binary", v, err)
 	}
 }
 
-// ToURIRef returns a url.URL value, decoding from string if necessary.
-func (v Value) ToURIRef() (url.URL, error) {
-	switch x := v.v.(type) {
-	case url.URL:
-		return x, nil
+// ToURL returns a *url.URL value, parsing from string if necessary.
+func ToURL(v interface{}) (*url.URL, error) {
+	v, err := Validate(v)
+	switch v := v.(type) {
+	case *url.URL:
+		return v, nil
 	case string:
-		u, err := url.Parse(x)
+		u, err := url.Parse(v)
 		if err != nil {
-			return url.URL{}, err
+			return nil, err
 		}
-		return *u, nil
+		return u, nil
 	default:
-		return url.URL{}, typeErr("URI-Reference", x)
+		return nil, convertErr("URL", v, err)
 	}
 }
 
-// ToTimestamp returns a Timestamp value, decoding from string if necessary.
-func (v Value) ToTimestamp() (time.Time, error) {
-	switch x := v.v.(type) {
+// ToTime returns a time.Time value, parsing from RFC3339 string if necessary.
+func ToTime(v interface{}) (time.Time, error) {
+	v, err := Validate(v)
+	switch v := v.(type) {
 	case time.Time:
-		return x, nil
+		return v, nil
 	case string:
-		ts, err := time.Parse(time.RFC3339Nano, x)
+		ts, err := time.Parse(time.RFC3339Nano, v)
 		if err != nil {
 			return time.Time{}, err
 		}
 		return ts, nil
 	default:
-		return time.Time{}, typeErr("Timestamp", x)
+		return time.Time{}, convertErr("Time", v, err)
 	}
 }
 
-func rangeErr(v interface{}) error {
-	return fmt.Errorf("%v is out of range for Integer", v)
-}
-
-func typeErr(name string, v interface{}) error {
+func convertErr(name string, v interface{}, err error) error {
```

Comment:
I like this, I wonder if we need a followup to add the IsConvertErr methods like k8s uses to see the type of error. like NotFoundError

Reply:
We could add a concrete
   type ConvertError string
and return that from convertErr, users can use standard type casts or type switches to detect. I prefer that to an IsBlah function for cases where we can control the error type.
We can do the function as well if you have a strong preference: 
    func IsConvertErr(err e) bool { _, ok := e.(ConvertError); return ok }